stustison/ddf
# Credential flow

Adds support for credential flow.

Test steps:
1. Build and startup this branch of DDF.
1. Startup keycloak by following this guide: https://connexta.atlassian.net/wiki/spaces/DDF/pages/702939155/Connect+Keycloak+to+DDF+using+OIDC.
1. Install DDF through the Admin UI. On the `SAML | OIDC` configuration page, make sure to configure the OIDC section according to the keycloak install guide. Note that the only different setting by default is the `secret`.
1. In the admin console, set the `Web Context Policy Manager` configuration's `Authentication Types` field to `/=OIDC|GUEST` & `/admin=SAML|IDP|BASIC`.
1. Attempt to login to Intrigue in an incognito browser. When redirected to the keycloak login page, enter in `localhost:localhost`. You should be logged in normally.
1. Verify you can re-log into Intrigue (try opening in a new tab) without having to authenticate against keycloak again.
1. Attempt to logout of Intrigue, and verify that once logged out you can no longer access Intrigue without authenticating again.
1. In the admin console, set the `OIDC Handler Configuration` configuration's `Default Response Type` field to `id_token`.
1. Attempt to login to Intrigue in an incognito browser. When redirected to the keycloak login page, enter in `localhost:localhost`. You should be logged in normally.
1. Verify you can re-log into Intrigue (try opening in a new tab) without having to authenticate against keycloak again.
1. Attempt to logout of Intrigue, and verify that once logged out you can no longer access Intrigue without authenticating again.
1. Put a breakpoint on line 180 in the `WebSsoFilter`: `if (result != null) {`, and start debugging;
1. In postman, make a POST request to `http://localhost:8080/auth/realms/master/protocol/openid-connect/token`
    1. Authorization: Basic Auth
        1. Username: `ddf-client`
        1. Password (secret): `{client_secret}`
    1. Headers:
        1. `Accept:application/json`
        1. `Cache-Control:no-cache`
        1. `Content-Type:application/x-www-form-urlencoded`
    1. Body:
        1. `grant_type:client_credentials`
1. Verify that the response is a 200 OK and contains an access token. Copy the value of the access token field for later use.
1. In postman, make a GET request to `https://localhost:8993/search/catalog`.
1. The breakpoint should get hit. Verify that the result is of source `karaf-GuestHandler`. Continue in the debugger.
1. In postman, verify that the response has a 200 OK status code.
1. In postman, make a GET request to `https://localhost:8993/search/catalog?access_token={copied_access_token}`.
1. The breakpoint should get hit. Verify that the result is of source `oidc-OidcHandler`. Continue in the debugger.
1. In postman, verify that the response has a 200 OK status code.
1. In postman, make a GET request to `https://localhost:8993/search/catalog`.
    1. Authorization: Bearer token
        1. Token: `{copied_access_token}`
1. The breakpoint should get hit. Verify that the result is of source `oidc-OidcHandler`. Continue in the debugger.
1. In postman, verify that the response has a 200 OK status code.
